File: Exercises/rb101-rb109_small_problems/easy_6/cute_angles.rb
# frozen_string_literal: true

=begin

# Problem

input: float
output: string

Rules:
-Explicit:
  - floating point input represents an angle between 0 and 360 degrees
  - output is a string that represents the angle in degrees, minutes and seconds
  - a degree has 60 minutes
  - a minute has 60 seconds

=end

DEGREE = "\xC2\xB0"

def convert(input)
  (input.divmod(1)[1] * 60).round(2)
end

def dms(angle)
  angle.negative? || angle > 360 ? degree = angle.modulo(360) : degree = angle
  minute = convert(degree)
  second = convert(minute)
  format(%(#{degree.truncate}#{DEGREE}%02d'%02d"), minute, second)
end

puts dms(30)        == %(30°00'00")
puts dms(76.73)     == %(76°43'48")
puts dms(254.6)     == %(254°36'00")
puts dms(93.034773) == %(93°02'05")
puts dms(0)         == %(0°00'00")
puts dms(360)       == %(360°00'00") || dms(360) == %(0°00'00")
puts dms(400) == %(40°00'00")
puts dms(-40) == %(320°00'00")
puts dms(-420) == %(300°00'00")

File: js210/exercises/easy_4/cute_angles.js
const DEGREE = "\xB0";
const MINUTES_PER_DEGREE = 60;
const SECONDS_PER_MINUTE = 60;

function dms(angle) {
  if (angle < 0) {
    angle = convertNegativeAngle(angle);
  }

  const degrees = formatNumber(Math.floor(angle));
  const hours = Math.floor((angle - degrees) * MINUTES_PER_DEGREE);
  const minutes = Math.floor(
    (angle % 1) * MINUTES_PER_DEGREE % 1 * SECONDS_PER_MINUTE
  );

  return `${formatNumber(degrees)}${DEGREE}${formatNumber(hours)}'${formatNumber(minutes)}"`;
}

function convertNegativeAngle(number) {
  while (number < 0) {
    number = 360 + number;
  }

  return number;
}

function formatNumber(number) {
  number = String(number);

  if (number.length < 2) {
    return `0${number}`;
  } else {
    return `${number}`;
  }
}

console.log(dms(30));           // 30°00'00"
console.log(dms(76.73));        // 76°43'48"
console.log(dms(254.6));        // 254°35'59"
console.log(dms(93.034773));    // 93°02'05"
console.log(dms(0));            // 0°00'00"
console.log(dms(360));          // 360°00'00" or 0°00'00"
console.log(dms(-1));
console.log(dms(-40));
console.log(dms(-420));
